forgerock-auth/src/androidTest/java/org/forgerock/android/auth/callback/DeviceBindingListAndUnbind.java

```diff
@@ -109,27 +111,140 @@ public void onException(Exception e) {
+        final int[] loginSuccess = {0};
+
+        NodeListenerFuture<FRSession> nodeListenerFuture = new DeviceSigningVerifierNodeListener(context, "default")
+        {
+            final NodeListener<FRSession> nodeListener = this;
+
+            @Override
+            public void onCallbackReceived(Node node)
+            {
+                if (node.getCallback(DeviceSigningVerifierCallback.class) != null) {
+                    DeviceSigningVerifierCallback callback = node.getCallback(DeviceSigningVerifierCallback.class);
+
+                    callback.sign(context, new FRListener<Void>() {
+                        @Override
+                        public void onSuccess(Void result) {
+                            node.next(context, nodeListener);
+                        }
+
+                        @Override
+                        public void onException(Exception e) {
+                            // Signing of the challenge has failed unexpectedly...
+                            Assertions.fail(e.getMessage());
+                        }
+                    });
+
+                    return;
+                }
+                if (node.getCallback(TextOutputCallback.class) != null) {
+                    TextOutputCallback textOutputCallback = node.getCallback(TextOutputCallback.class);
+                    assertThat(textOutputCallback.getMessage()).isEqualTo("Success");
+                    loginSuccess[0]++;
+
+                    node.next(context, nodeListener);
+                    return;
+                }
+
+                super.onCallbackReceived(node);
+            }
+        };
+
+        FRSession.authenticate(context, TREE, nodeListenerFuture);
+        Assert.assertNotNull(nodeListenerFuture.get());
+
+        // Ensure that the journey finishes with success
+        Assert.assertNotNull(FRSession.getCurrentSession());
+        Assert.assertNotNull(FRSession.getCurrentSession().getSessionToken());
+    }
+
     @Test
     public void testListAndDeleteKeys() throws ExecutionException, InterruptedException, IOException, ApiException {
+        final int[] deletionFailure = {0};
+        final int[] deletionSuccess = {0};
         FRUserKeys userKeys = new FRUserKeys(context);
         assertThat(userKeys.loadAll().size()).isEqualTo(2);
 
         // Assert that the keys are indeed associated with the correct user
         assertThat(userKeys.loadAll().get(0).getUserName()).isEqualTo(USERNAME);
         assertThat(userKeys.loadAll().get(1).getUserName()).isEqualTo(USERNAME);
 
-        // Delete one of the keys and confirm it was deleted
-        FRListenerFuture<Void> future = new FRListenerFuture<>();
+        // Attempt to remove a key without being authenticated with it should fail
+        FRListenerFuture<Void> future = new FRListenerFuture<Void>();
         userKeys.delete(userKeys.loadAll().get(0), false, future);
-        future.get();
+        try {
+            future.get();
+            deletionSuccess[0]++;
```

Comment:
You can just use fail() function, with fail(), it throws exception if the code reach that point.

Reply:
Fixed.